### PR TITLE
Create bank snapshots

### LINKED ISSUE #2475: Fullnode startup is slow
#### Problem

Currently, starting a fullnode queries the cluster for the full ledger and replays it.

#### Proposed Solution

The fullnode should first check for a recent ledger state snapshot locally and if unavailable, query another validator for one.

### PR DESCRIPTION
**Pull request by sambley**
#### Problem
Full node startup is slow when many transactions have been processed as this involved querying for the full ledger and replaying it.

#### Summary of Changes
Serialize bank state into a snapshot and try to restore from that on boot.

Fixes #2475

### CODE REVIEW

**Comment by sambley** [comment]
@sakridge, could you check on how this is looking

File: runtime/src/bloom.rs [hunk 1]
```diff
@@ -16,7 +16,8 @@ pub trait BloomHashIndex {
 #[derive(Serialize, Deserialize, Default, Clone, Debug, PartialEq)]
 pub struct Bloom<T: BloomHashIndex> {
     pub keys: Vec<u64>,
-    pub bits: BitVec<u8>,
+    pub bits: BitVec<u64>,
+    num_bits_set: u64,
```
Comment (sakridge) [inline runtime/src/bloom.rs]: Do we use this anywhere?

Comment (sambley) [reply runtime/src/bloom.rs]: its being used in crds gossip push and pull

**Comment by aeyakovenko** [comment]
@sambley, can you take a look at #3613. Index can be recovered from storage.  We just need the last root from blocktree

**Comment by mvines** [comment]
Is this PR still being worked on?

**Comment by sambley** [comment]
> Is this PR still being worked on?

@mvines, yes I am working on this one.

**Comment by aeyakovenko** [comment]
@sambley I think scan + getting the root from blocktree should get you there

**Comment by sambley** [comment]
> @sambley I think scan + getting the root from blocktree should get you there

@aeyakovenko, currently the frozen set of banks (32) are serialized (added/removed as they are pruned) with the serialized data containing the path to the append vec and when deserializing the accounts is restored back by pointing to the appropriate append vec. When you are referring to scan + getting root from blocktree, are you referring to the above or a different mechanism.

**Comment by aeyakovenko** [comment]
@sambley 

https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L196

There are examples of usage in accounts.rs. But basically the appendvec stores a single fork, so booting the checkpoint just means scanning all the vecs and reconstructing the index based on write_version.

**Comment by sambley** [comment]
> @sambley
> 
> https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L196
> 
> There are examples of usage in accounts.rs. But basically the appendvec stores a single fork, so booting the checkpoint just means scanning all the vecs and reconstructing the index based on write_version.

@aeyakovenko, currently AccountsIndex get serialized / deserialized and should restore back the index on deserialize. The bank snapshot are added / removed with below logic.
https://github.com/solana-labs/solana/pull/3671/files#diff-0e517bfb897cd6ee46d83d239efb1cb7R131

Do you see any issue with this?

**Comment by aeyakovenko** [comment]
@sambley, it doesn't need to be.  if we recover from scanning the storage, we can also GC it in that step and recreate the index.  Serializing the index adds another level of complexity, since it's hard to make sure the index is in sync with storage.

File: runtime/Cargo.toml [hunk 2]
```diff
@@ -20,7 +20,7 @@ log = "0.4.2"
 memmap = "0.6.2"
 rand = "0.6.5"
 rayon = "1.0.0"
-serde = "1.0.88"
+serde = { version = "1.0.88", features = ["rc"] }
```
Comment (rob-solana) [inline runtime/Cargo.toml]: this looks like it might be a really nasty thing to turn on...

Comment (sambley) [reply runtime/Cargo.toml]: I will work on serializing / deserializing with out having to turn on this feature.

**Comment by aeyakovenko** [comment]
@sambley so my goal with using AppendVec data directly without a saved index is that the bank can recover from a killed process, and the index can be fresh, basically garbage collected from boot.

**Comment by sambley** [comment]
> @sambley so my goal with using AppendVec data directly without a saved index is that the bank can recover from a killed process, and the index can be fresh, basically garbage collected from boot.

Yes, I will work on that as well to not have to serialize the index and instead recover from the appendvec. Still looking into an issue with the continuation once the bank snapshots are restored.

**Comment by aeyakovenko** [comment]
@sambley @mvines so once we have local snapshots, we need a way to download them from the network

**Comment by mvines** [comment]
I'm itching to nuke our rsync usage from orbit, I've just been waiting for snapshots to progress further since I figured our rsync of the genesis block essentially could look like the degenerate case of downloading a snapshot.   I'll gleefully take on the work of plumbing snapshot downloads once they're making it on disk locally somewhere

**Comment by sambley** [comment]
@aeyakovenko, noticing that the snapshot size quickly grows and StatusCache is the major contributor to the size of snapshot (other than the accounts db which is stored separately). Does the entire cache need to be serialized or can it be optimized?

**Comment by aeyakovenko** [comment]
Only the signatures per block are needed

**Comment by rob-solana** [comment]
sorry, Sathish: I fat-fingered the merge button

**Pull request merged**